cmd/copy.go

```diff
@@ -948,6 +1025,136 @@ func (cca *cookedCopyCmdArgs) processCopyJobPartOrders() (err error) {
+	switch location {
+	case common.ELocation.Local():
+		return nil, nil
+	case common.ELocation.Blob():
+		p, err = createBlobPipeline(ctx, credential)
+	case common.ELocation.File():
+		p, err = createFilePipeline(ctx, credential)
+	case common.ELocation.BlobFS():
+		p, err = createBlobFSPipeline(ctx, credential)
+	default:
+		err = fmt.Errorf("can't produce new pipeline for location %s", location)
+	}
+
+	return
+}
+
+func trimWildcards(path string) string {
+	if strings.Index(path, "*") == -1 {
+		return path
+	}
+
+	firstWCIndex := strings.Index(path, "*")
+	result := replacePathSeparators(path[:firstWCIndex])
+	lastSepIndex := strings.LastIndex(result, "/")
+	result = result[:lastSepIndex+1]
+
+	return result
+}
+
+func (cca *cookedCopyCmdArgs) findObject(source bool, dstIsDir bool, object storedObject) (relativePath string) {
+	var pathEncodeRules = func(path string) string {
+		loc := common.ELocation.Unknown()
+
+		if source {
+			loc = cca.fromTo.From()
+		} else {
+			loc = cca.fromTo.To()
+		}
+		pathParts := strings.Split(path, "/")
+
+		// Encode disallowed characters on windows.
+		// TODO: Inform tests of this
+		if loc == common.ELocation.Local() && runtime.GOOS == "windows" {
+			invalidChars := `<>\/:"|?*` + string(0x00)
+
+			for _, c := range strings.Split(invalidChars, "") {
+				for k, p := range pathParts {
+					pathParts[k] = strings.ReplaceAll(p, c, url.PathEscape(c))
+				}
+			}
+		} else if loc != common.ELocation.Local() {
+			for k, p := range pathParts {
+				pathParts[k] = url.PathEscape(p)
+			}
+		}
+
+		path = strings.Join(pathParts, "/")
+		return path
+	}
+
+	// source is a EXACT path to the file.
+	if object.relativePath == "" {
+		// If we're finding an object from the source, it returns "" if it's already got it.
+		// If we're finding an object on the destination and we get "", we need to hand it the object name (if it's pointing to a folder)
+		if source {
+			relativePath = ""
+		} else {
+			if dstIsDir {
+				relativePath = "/" + object.name
+			} else {
+				relativePath = ""
+			}
+		}
+
+		return pathEncodeRules(relativePath)
+	}
+
+	// If it's out here, the source is a folder (or wildcard) of some kind.
+
+	relativePath = "/" + strings.Replace(object.relativePath, common.OS_PATH_SEPARATOR, common.AZCOPY_PATH_SEPARATOR_STRING, -1)
+
+	if !source && !strings.Contains(cca.source, "*") {
+		// We ONLY need to do this adjustment to the destination.
+		// The source SAS has already been removed. No need to convert it to a URL or whatever.
+		// Save to a directory
+		relativePath = "/" + filepath.Base(cca.source) + relativePath
+	}
+
+	return pathEncodeRules(relativePath)
+}
+
+// Initialize the modular filters outside of copy to increase readability.
+func (cca *cookedCopyCmdArgs) initModularFilters() []objectFilter {
+	filters := make([]objectFilter, 0) // same as []objectFilter{} under the hood
```

Comment:
What's the plan for incorporating the new filters from commit 9008a3b?  (Asking because we need to make sure we don't forget?  Don't want to have them get unhooked from the codepaths, just because they're new and we are refactoing)  Maybe could make a PBI to hook them in in next sprint?

Reply:
Ping me once they get merged in and I'll merge them into this branch.

Reply:
Good question. Let me merge them into this branch and include them now. No reason they _can't_ be included yet.